src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,179 @@
+    m_spend_pubkey = CPubKey{m_spend_seckey.GetPubKey()};
+}
+
+void Recipient::ComputeECDHSharedSecret(const CPubKey& sender_public_key, const std::vector<COutPoint>& tx_outpoints)
+{
+    const auto& outpoint_hash = HashOutpoints(tx_outpoints);
+    auto tweaked_scan_seckey = m_scan_seckey.MultiplyTweak(outpoint_hash.begin());
+    CPubKey result = tweaked_scan_seckey.SilentPaymentECDH(sender_public_key);
+
+    m_ecdh_pubkey = result;
+    assert(m_ecdh_pubkey.IsValid());
+}
+
+std::pair<CKey,CPubKey> Recipient::CreateOutput(const uint32_t output_index) const
+{
+    HashWriter h;
+    h.write(Span{m_ecdh_pubkey});
+    unsigned char num[4];
+    WriteBE32(num, output_index);
+    h << num;
+    uint256 shared_secret = h.GetSHA256();
+
+    const auto& result_pubkey{m_spend_pubkey.AddTweak(shared_secret.begin())};
+    const auto& result_seckey{m_spend_seckey.AddTweak(shared_secret.begin())};
+
+    return {result_seckey, result_pubkey};
+}
+
+std::pair<CPubKey,CPubKey> Recipient::GetAddress() const
+{
+    return {m_scan_pubkey, m_spend_pubkey};
+}
+
+std::vector<CKey> Recipient::ScanTxOutputs(std::vector<XOnlyPubKey> output_pub_keys)
+{
+    // Because a sender can create multiple outputs for us, we first check the outputs vector for an output with
+    // output index 0. If we find it, we remove it from the vector and then iterate over the vector again looking for
+    // an output with index 1, and so on until one of the following happens:
+    //
+    //     1. We have determined all outputs belong to us (the vector is empty)
+    //     2. We have passed over the vector and found no outputs belonging to us
+    //
+
+    // The ECDH shared secret must be set before we start scanning
+    assert(m_ecdh_pubkey.IsValid());
+
+    bool removed;
+    uint32_t output_index{0};
+    std::vector<CKey> raw_tr_keys;
+    do {
+        // We haven't removed anything yet on this pass and if we don't remove anything, we didn't find
+        // any silent payment outputs and should stop checking
+        removed = false;
+        std::pair<CKey, CPubKey> tweakResult = CreateOutput(output_index);
+        const CKey& silent_payment_priv_key = tweakResult.first;
+        const XOnlyPubKey& silent_payment_pub_key = XOnlyPubKey{tweakResult.second};
+        output_pub_keys.erase(std::remove_if(output_pub_keys.begin(), output_pub_keys.end(), [&](auto outputPubKey) {
+            if (silent_payment_pub_key == outputPubKey) {
+                // Since we found an output, we need to increment the output index and check the vector again
+                raw_tr_keys.emplace_back(silent_payment_priv_key);
+                removed = true;
+                output_index++;
+                // Return true so that this output pubkey is removed the from vector and not checked again
+                return true;
+            }
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && removed);
+    return raw_tr_keys;
```

Comment:
This shouldn't return private keys, it should only return the tweak portion of the private key. This also means `Recipient` doesn't need to have the spend private key.